In [2]:
#importing some libraries
import pandas as pd
import matplotlib.pyplot as ptl
%matplotlib inline
import seaborn as sns

In [3]:
#converting text data to dataframe
df = pd.read_csv('47965_1_120_20200701_125722_dat.txt', sep="|", header=None)
df.head()
df.tail()

c:\users\arnab\appdata\local\programs\python\python37\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


0                            1         2       3         4   \
56643  593720            GLEAM FABMAT LTD.  20190326    5.26  1.95e+06   
56644  593720            GLEAM FABMAT LTD.  20190327    5.78    330000   
56645  593720            GLEAM FABMAT LTD.  20190329    6.35     10000   
56646  594045  HUMMING BIRD EDUCATION LTD.  20190328     132      7000   
56647  594045  HUMMING BIRD EDUCATION LTD.  20190329  132.05      8000   

           5        6    7    8    9    10   11   12   13   14   15   16   17  \
56643    5.05  52.6947  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
56644    5.46   57.904  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
56645    6.35  63.6143  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
56646  132.98   80.916  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
56647  132.34  80.9466  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

        18          19  
56643  NaN  1.0018e+07  
56644  NaN  1.0018e+07  
56645  NaN  1.0018e+07  
56646  NaN      613000  
56647  NaN      613000

In [4]:
#Correcting the column names
df.rename(columns=df.iloc[0], inplace = True)
df.drop([0], inplace = True)
df.head()

co_code     company_name co_stkdate bse_closing_price bse_traded_qty  \
1      11  20 MICRONS LTD.   20190301           37.1500           2286   
2      11  20 MICRONS LTD.   20190305           39.7500           1976   
3      11  20 MICRONS LTD.   20190306           41.0000           2528   
4      11  20 MICRONS LTD.   20190307           39.6000           5649   
5      11  20 MICRONS LTD.   20190308           38.4000           1783   

  bse_avg_rate bse_market_cap nse_opening_price nse_high_price nse_low_price  \
1      37.4900      1310.8935           37.3000        38.0000       37.1000   
2      39.0700      1402.6385           38.4000        40.0000       37.8500   
3      40.8600      1446.7466           40.4000        41.4500       39.7500   
4      39.9000      1397.3455           41.4000        41.4000       39.0000   
5      38.6800      1355.0017           39.0000        40.1000       38.1000   

  nse_closing_price nse_returns nse_traded_qty nse_traded_val nse_avg_rate  \
1           37.8000      1.0300          30685         1.2000      37.6200   
2           39.8500      1.0500          32809         1.3000      39.0500   
3           40.5500      1.0200          38445         1.6000      40.7000   
4           39.4500      0.9700          20465         0.8000      40.2500   
5           38.4000      0.9700          30713         1.2000      39.0600   

  nse_market_cap  nse_pe  nse_pb nse_bv_per_share equity_shares_on_stkdate  
1      1333.8298  6.8192  0.8971          42.1379            35286502.0000  
2      1406.1671  7.1890  0.9457          42.1379            35286502.0000  
3      1430.8677  7.3153  0.9623          42.1379            35286502.0000  
4      1392.0525  7.1168  0.9362          42.1379            35286502.0000  
5      1355.0017  6.9274  0.9113          42.1379            35286502.0000

### Inspecting the data

In [5]:
df.shape #56647 rows, 20 columns

(56647, 20)

In [6]:
df.describe()

co_code         company_name co_stkdate bse_closing_price  \
count    56647                56647      56647             53154   
unique    3868                 3868         38             20432   
top      40158  5PAISA CAPITAL LTD.   20190313            0.1900   
freq        19                   19       1755               211   

       bse_traded_qty bse_avg_rate  bse_market_cap  nse_opening_price  \
count           53154        53154         53154.0            32137.0   
unique          28792        31493         47629.0            14278.0   
top               100       0.1900             0.0                0.3   
freq              478          190            37.0               50.0   

        nse_high_price  nse_low_price  nse_closing_price nse_returns  \
count         32137.00       32137.00            32137.0       32137   
unique        14221.00       14315.00            16099.0          90   
top               0.35           0.25                0.3      1.0000   
freq             59.00          59.00               69.0        3855   

        nse_traded_qty nse_traded_val  nse_avg_rate  nse_market_cap   nse_pe  \
count          32137.0          32137      32137.00         32137.0  23470.0   
unique         28353.0           6720      24701.00         28634.0  21753.0   
top             4000.0         0.0000          1.25             0.0      0.0   
freq             165.0           1342         29.00            38.0     38.0   

         nse_pb nse_bv_per_share equity_shares_on_stkdate  
count   29311.0            32099                    56609  
unique  23431.0             2030                     3974  
top         0.0           0.0000             5000000.0000  
freq       38.0              171                      138

In [7]:
df.info()
# PROBLEM - all columns are string, have to convert them to int

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56647 entries, 1 to 56647
Data columns (total 20 columns):
co_code                     56647 non-null object
company_name                56647 non-null object
co_stkdate                  56647 non-null object
bse_closing_price           53154 non-null object
bse_traded_qty              53154 non-null object
bse_avg_rate                53154 non-null object
bse_market_cap              53154 non-null object
nse_opening_price           32137 non-null object
nse_high_price              32137 non-null object
nse_low_price               32137 non-null object
nse_closing_price           32137 non-null object
nse_returns                 32137 non-null object
nse_traded_qty              32137 non-null object
nse_traded_val              32137 non-null object
nse_avg_rate                32137 non-null object
nse_market_cap              32137 non-null object
nse_pe                      23470 non-null object
nse_pb                      29311 non-nul

In [8]:
#renaming the column co_stkdate
df.rename({'co_stkdate': 'date'}, axis=1, inplace=True)
df.head()

co_code     company_name      date bse_closing_price bse_traded_qty  \
1      11  20 MICRONS LTD.  20190301           37.1500           2286   
2      11  20 MICRONS LTD.  20190305           39.7500           1976   
3      11  20 MICRONS LTD.  20190306           41.0000           2528   
4      11  20 MICRONS LTD.  20190307           39.6000           5649   
5      11  20 MICRONS LTD.  20190308           38.4000           1783   

  bse_avg_rate bse_market_cap nse_opening_price nse_high_price nse_low_price  \
1      37.4900      1310.8935           37.3000        38.0000       37.1000   
2      39.0700      1402.6385           38.4000        40.0000       37.8500   
3      40.8600      1446.7466           40.4000        41.4500       39.7500   
4      39.9000      1397.3455           41.4000        41.4000       39.0000   
5      38.6800      1355.0017           39.0000        40.1000       38.1000   

  nse_closing_price nse_returns nse_traded_qty nse_traded_val nse_avg_rate  \
1           37.8000      1.0300          30685         1.2000      37.6200   
2           39.8500      1.0500          32809         1.3000      39.0500   
3           40.5500      1.0200          38445         1.6000      40.7000   
4           39.4500      0.9700          20465         0.8000      40.2500   
5           38.4000      0.9700          30713         1.2000      39.0600   

  nse_market_cap  nse_pe  nse_pb nse_bv_per_share equity_shares_on_stkdate  
1      1333.8298  6.8192  0.8971          42.1379            35286502.0000  
2      1406.1671  7.1890  0.9457          42.1379            35286502.0000  
3      1430.8677  7.3153  0.9623          42.1379            35286502.0000  
4      1392.0525  7.1168  0.9362          42.1379            35286502.0000  
5      1355.0017  6.9274  0.9113          42.1379            35286502.0000

In [9]:
df.isnull().sum() #many columns have null values

co_code                         0
company_name                    0
date                            0
bse_closing_price            3493
bse_traded_qty               3493
bse_avg_rate                 3493
bse_market_cap               3493
nse_opening_price           24510
nse_high_price              24510
nse_low_price               24510
nse_closing_price           24510
nse_returns                 24510
nse_traded_qty              24510
nse_traded_val              24510
nse_avg_rate                24510
nse_market_cap              24510
nse_pe                      33177
nse_pb                      27336
nse_bv_per_share            24548
equity_shares_on_stkdate       38
dtype: int64

In [11]:
df.tail()

co_code                 company_name      date bse_closing_price  \
56643  593720            GLEAM FABMAT LTD.  20190326              5.26   
56644  593720            GLEAM FABMAT LTD.  20190327              5.78   
56645  593720            GLEAM FABMAT LTD.  20190329              6.35   
56646  594045  HUMMING BIRD EDUCATION LTD.  20190328               132   
56647  594045  HUMMING BIRD EDUCATION LTD.  20190329            132.05   

      bse_traded_qty bse_avg_rate bse_market_cap nse_opening_price  \
56643       1.95e+06         5.05        52.6947               NaN   
56644         330000         5.46         57.904               NaN   
56645          10000         6.35        63.6143               NaN   
56646           7000       132.98         80.916               NaN   
56647           8000       132.34        80.9466               NaN   

      nse_high_price nse_low_price nse_closing_price nse_returns  \
56643            NaN           NaN               NaN         NaN   
56644            NaN           NaN               NaN         NaN   
56645            NaN           NaN               NaN         NaN   
56646            NaN           NaN               NaN         NaN   
56647            NaN           NaN               NaN         NaN   

      nse_traded_qty nse_traded_val nse_avg_rate nse_market_cap nse_pe nse_pb  \
56643            NaN            NaN          NaN            NaN    NaN    NaN   
56644            NaN            NaN          NaN            NaN    NaN    NaN   
56645            NaN            NaN          NaN            NaN    NaN    NaN   
56646            NaN            NaN          NaN            NaN    NaN    NaN   
56647            NaN            NaN          NaN            NaN    NaN    NaN   

      nse_bv_per_share equity_shares_on_stkdate  
56643              NaN               1.0018e+07  
56644              NaN               1.0018e+07  
56645              NaN               1.0018e+07  
56646              NaN                   613000  
56647              NaN                   613000

In [12]:
print(len(df.company_name.unique())) #there are 3868 companies

3868


In [13]:
df_company = pd.read_excel('companies.xlsx')
df_company.head()

KZ Index  co_code                     sa_company_name  \
0 -333.344377   175908        MARATHON NEXTGEN REALTY LTD.   
1 -300.207235   318534            WELSPUN ENTERPRISES LTD.   
2 -192.347057   322563          FIRSTSOURCE SOLUTIONS LTD.   
3 -133.326392   220218  SHERVANI INDUSTRIAL SYNDICATE LTD.   
4 -107.045089   212051                    INDIANIVESH LTD.   

                           Ownership  Quartile  
0                     Marathon group         1  
1                      Welspun Group         1  
2              RPG Enterprises Group         1  
3                   Private (Indian)         1  
4  Industrial Investment Trust Group         1

In [14]:
df_company.columns

Index(['KZ Index', 'co_code', 'sa_company_name', 'Ownership', 'Quartile'], dtype='object')

In [15]:
df_merged = pd.merge(df, df_company, on='co_code')
df_merged

co_code                     company_name      date bse_closing_price  \
0     192422             RAM RATNA WIRES LTD.  20190301              96.9   
1     192422             RAM RATNA WIRES LTD.  20190305             107.6   
2     192422             RAM RATNA WIRES LTD.  20190306            102.35   
3     192422             RAM RATNA WIRES LTD.  20190307             102.9   
4     192422             RAM RATNA WIRES LTD.  20190308               104   
...      ...                              ...       ...               ...   
4204  571292  MADHYA PRADESH TODAY MEDIA LTD.  20190320               NaN   
4205  571292  MADHYA PRADESH TODAY MEDIA LTD.  20190322               NaN   
4206  571292  MADHYA PRADESH TODAY MEDIA LTD.  20190325               NaN   
4207  571292  MADHYA PRADESH TODAY MEDIA LTD.  20190326               NaN   
4208  571292  MADHYA PRADESH TODAY MEDIA LTD.  20190329               NaN   

     bse_traded_qty bse_avg_rate bse_market_cap nse_opening_price  \
0              5162        97.05         2131.8               NaN   
1             18902       102.34         2367.2               NaN   
2             15058       105.47         2251.7               NaN   
3             11899       102.22         2263.8               NaN   
4               360       104.43           2288               NaN   
...             ...          ...            ...               ...   
4204            NaN          NaN            NaN              38.3   
4205            NaN          NaN            NaN              36.4   
4206            NaN          NaN            NaN              42.1   
4207            NaN          NaN            NaN             40.25   
4208            NaN          NaN            NaN              38.2   

     nse_high_price nse_low_price  ... nse_avg_rate nse_market_cap  nse_pe  \
0               NaN           NaN  ...          NaN            NaN     NaN   
1               NaN           NaN  ...          NaN            NaN     NaN   
2               NaN           NaN  ...          NaN            NaN     NaN   
3               NaN           NaN  ...          NaN            NaN     NaN   
4               NaN           NaN  ...          NaN            NaN     NaN   
...             ...           ...  ...          ...            ...     ...   
4204           38.3          38.3  ...         38.3        174.954  3.7463   
4205          40.15          36.4  ...        38.54        183.405  3.9273   
4206           42.1          42.1  ...         42.1        192.313   4.118   
4207          40.25          40.1  ...        40.22        183.405  3.9273   
4208           38.2         38.15  ...        38.17        174.269  3.7317   

      nse_pb nse_bv_per_share equity_shares_on_stkdate  KZ Index  \
0        NaN              NaN                  2.2e+07 -0.335171   
1        NaN              NaN                  2.2e+07 -0.335171   
2        NaN              NaN                  2.2e+07 -0.335171   
3        NaN              NaN                  2.2e+07 -0.335171   
4        NaN              NaN                  2.2e+07 -0.335171   
...      ...              ...                      ...       ...   
4204  0.7092          54.0061                4.568e+06 -1.284226   
4205  0.7434          54.0061                4.568e+06 -1.284226   
4206  0.7795          54.0061                4.568e+06 -1.284226   
4207  0.7434          54.0061                4.568e+06 -1.284226   
4208  0.7064          54.0061                4.568e+06 -1.284226   

                      sa_company_name         Ownership Quartile  
0                RAM RATNA WIRES LTD.   Ram Ratna Group        3  
1                RAM RATNA WIRES LTD.   Ram Ratna Group        3  
2                RAM RATNA WIRES LTD.   Ram Ratna Group        3  
3                RAM RATNA WIRES LTD.   Ram Ratna Group        3  
4                RAM RATNA WIRES LTD.   Ram Ratna Group        3  
...                               ...               ...      ...  
4204  MADHYA PRADESH TODAY M

In [16]:
len(df_merged['company_name'].unique())

227

In [17]:
df_merged.drop(['sa_company_name'], axis=1,inplace = True)

In [18]:
df_merged.head()

co_code          company_name      date bse_closing_price bse_traded_qty  \
0  192422  RAM RATNA WIRES LTD.  20190301              96.9           5162   
1  192422  RAM RATNA WIRES LTD.  20190305             107.6          18902   
2  192422  RAM RATNA WIRES LTD.  20190306            102.35          15058   
3  192422  RAM RATNA WIRES LTD.  20190307             102.9          11899   
4  192422  RAM RATNA WIRES LTD.  20190308               104            360   

  bse_avg_rate bse_market_cap nse_opening_price nse_high_price nse_low_price  \
0        97.05         2131.8               NaN            NaN           NaN   
1       102.34         2367.2               NaN            NaN           NaN   
2       105.47         2251.7               NaN            NaN           NaN   
3       102.22         2263.8               NaN            NaN           NaN   
4       104.43           2288               NaN            NaN           NaN   

   ... nse_traded_val nse_avg_rate nse_market_cap nse_pe nse_pb  \
0  ...            NaN          NaN            NaN    NaN    NaN   
1  ...            NaN          NaN            NaN    NaN    NaN   
2  ...            NaN          NaN            NaN    NaN    NaN   
3  ...            NaN          NaN            NaN    NaN    NaN   
4  ...            NaN          NaN            NaN    NaN    NaN   

  nse_bv_per_share equity_shares_on_stkdate  KZ Index        Ownership  \
0              NaN                  2.2e+07 -0.335171  Ram Ratna Group   
1              NaN                  2.2e+07 -0.335171  Ram Ratna Group   
2              NaN                  2.2e+07 -0.335171  Ram Ratna Group   
3              NaN                  2.2e+07 -0.335171  Ram Ratna Group   
4              NaN                  2.2e+07 -0.335171  Ram Ratna Group   

  Quartile  
0        3  
1        3  
2        3  
3        3  
4        3  

[5 rows x 23 columns]

In [19]:
#Tried converting date to datetime using pandas but not working
#Looping through values of date first
#newDate = []
#for item in df_merged['date']:
    #item_v = str(item[0:3]) + '-' + str(item[4:5]) + '-' + str(item[6:7])
    #newDate.append(item_v)
#print(newDate)
#df_merged.loc[:, 'date'] = newDate

df_merged['date'] = pd.to_datetime(df_merged['date'], format = '%Y%m%d')
df_merged

co_code                     company_name       date bse_closing_price  \
0     192422             RAM RATNA WIRES LTD. 2019-03-01              96.9   
1     192422             RAM RATNA WIRES LTD. 2019-03-05             107.6   
2     192422             RAM RATNA WIRES LTD. 2019-03-06            102.35   
3     192422             RAM RATNA WIRES LTD. 2019-03-07             102.9   
4     192422             RAM RATNA WIRES LTD. 2019-03-08               104   
...      ...                              ...        ...               ...   
4204  571292  MADHYA PRADESH TODAY MEDIA LTD. 2019-03-20               NaN   
4205  571292  MADHYA PRADESH TODAY MEDIA LTD. 2019-03-22               NaN   
4206  571292  MADHYA PRADESH TODAY MEDIA LTD. 2019-03-25               NaN   
4207  571292  MADHYA PRADESH TODAY MEDIA LTD. 2019-03-26               NaN   
4208  571292  MADHYA PRADESH TODAY MEDIA LTD. 2019-03-29               NaN   

     bse_traded_qty bse_avg_rate bse_market_cap nse_opening_price  \
0              5162        97.05         2131.8               NaN   
1             18902       102.34         2367.2               NaN   
2             15058       105.47         2251.7               NaN   
3             11899       102.22         2263.8               NaN   
4               360       104.43           2288               NaN   
...             ...          ...            ...               ...   
4204            NaN          NaN            NaN              38.3   
4205            NaN          NaN            NaN              36.4   
4206            NaN          NaN            NaN              42.1   
4207            NaN          NaN            NaN             40.25   
4208            NaN          NaN            NaN              38.2   

     nse_high_price nse_low_price  ... nse_traded_val nse_avg_rate  \
0               NaN           NaN  ...            NaN          NaN   
1               NaN           NaN  ...            NaN          NaN   
2               NaN           NaN  ...            NaN          NaN   
3               NaN           NaN  ...            NaN          NaN   
4               NaN           NaN  ...            NaN          NaN   
...             ...           ...  ...            ...          ...   
4204           38.3          38.3  ...            0.1         38.3   
4205          40.15          36.4  ...            0.5        38.54   
4206           42.1          42.1  ...            0.1         42.1   
4207          40.25          40.1  ...            0.4        40.22   
4208           38.2         38.15  ...            0.2        38.17   

     nse_market_cap  nse_pe  nse_pb nse_bv_per_share equity_shares_on_stkdate  \
0               NaN     NaN     NaN              NaN                  2.2e+07   
1               NaN     NaN     NaN              NaN                  2.2e+07   
2               NaN     NaN     NaN              NaN                  2.2e+07   
3               NaN     NaN     NaN              NaN                  2.2e+07   
4               NaN     NaN     NaN              NaN                  2.2e+07   
...             ...     ...     ...              ...                      ...   
4204        174.954  3.7463  0.7092          54.0061                4.568e+06   
4205        183.405  3.9273  0.7434          54.0061                4.568e+06   
4206        192.313   4.118  0.7795          54.0061                4.568e+06   
4207        183.405  3.9273  0.7434          54.0061                4.568e+06   
4208        174.269  3.7317  0.7064          54.0061                4.568e+06   

      KZ Index         Ownership Quartile  
0    -0.335171   Ram Ratna Group        3  
1    -0.335171   Ram Ratna Group        3  
2    -0.335171   Ram Ratna Group        3  
3    -0.335171   Ram Ratna Group        3  
4    -0.335171   Ram Ratna Group        3  
...        ...               ...      ...  
4204 -1.284226  Private (Indian)        2  
4205 -1.284226  Private (Indian)        2  
4206 -1.284226  Private (Indi

In [20]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4209 entries, 0 to 4208
Data columns (total 23 columns):
co_code                     4209 non-null object
company_name                4209 non-null object
date                        4209 non-null datetime64[ns]
bse_closing_price           3974 non-null object
bse_traded_qty              3974 non-null object
bse_avg_rate                3974 non-null object
bse_market_cap              3974 non-null object
nse_opening_price           3144 non-null object
nse_high_price              3144 non-null object
nse_low_price               3144 non-null object
nse_closing_price           3144 non-null object
nse_returns                 3144 non-null object
nse_traded_qty              3144 non-null object
nse_traded_val              3144 non-null object
nse_avg_rate                3144 non-null object
nse_market_cap              3144 non-null object
nse_pe                      3030 non-null object
nse_pb                      3144 non-null object
nse

In [21]:
#converting other columns from string to float. Filling in 0.0 for null values for now
#for a in df.columns[3:]:
    #df.a = df.a.astype(float).fillna(0.0)
    
#above code didnt work so we do it automatically
df_merged = df_merged.infer_objects() 
print(df_merged.dtypes)

co_code                              int64
company_name                        object
date                        datetime64[ns]
bse_closing_price                  float64
bse_traded_qty                     float64
bse_avg_rate                       float64
bse_market_cap                     float64
nse_opening_price                  float64
nse_high_price                     float64
nse_low_price                      float64
nse_closing_price                  float64
nse_returns                        float64
nse_traded_qty                     float64
nse_traded_val                     float64
nse_avg_rate                       float64
nse_market_cap                     float64
nse_pe                             float64
nse_pb                             float64
nse_bv_per_share                   float64
equity_shares_on_stkdate           float64
KZ Index                           float64
Ownership                           object
Quartile                             int64
dtype: obje

In [22]:
df_merged.info()
#datatype conversion successful

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4209 entries, 0 to 4208
Data columns (total 23 columns):
co_code                     4209 non-null int64
company_name                4209 non-null object
date                        4209 non-null datetime64[ns]
bse_closing_price           3974 non-null float64
bse_traded_qty              3974 non-null float64
bse_avg_rate                3974 non-null float64
bse_market_cap              3974 non-null float64
nse_opening_price           3144 non-null float64
nse_high_price              3144 non-null float64
nse_low_price               3144 non-null float64
nse_closing_price           3144 non-null float64
nse_returns                 3144 non-null float64
nse_traded_qty              3144 non-null float64
nse_traded_val              3144 non-null float64
nse_avg_rate                3144 non-null float64
nse_market_cap              3144 non-null float64
nse_pe                      3030 non-null float64
nse_pb                      3144 non-nu

In [23]:
df_merged.head(10)

co_code          company_name       date  bse_closing_price  \
0   192422  RAM RATNA WIRES LTD. 2019-03-01              96.90   
1   192422  RAM RATNA WIRES LTD. 2019-03-05             107.60   
2   192422  RAM RATNA WIRES LTD. 2019-03-06             102.35   
3   192422  RAM RATNA WIRES LTD. 2019-03-07             102.90   
4   192422  RAM RATNA WIRES LTD. 2019-03-08             104.00   
5   192422  RAM RATNA WIRES LTD. 2019-03-11             105.50   
6   192422  RAM RATNA WIRES LTD. 2019-03-12             106.45   
7   192422  RAM RATNA WIRES LTD. 2019-03-13             106.80   
8   192422  RAM RATNA WIRES LTD. 2019-03-14             107.10   
9   192422  RAM RATNA WIRES LTD. 2019-03-15             105.50   

   bse_traded_qty  bse_avg_rate  bse_market_cap  nse_opening_price  \
0          5162.0         97.05          2131.8                NaN   
1         18902.0        102.34          2367.2                NaN   
2         15058.0        105.47          2251.7                NaN   
3         11899.0        102.22          2263.8                NaN   
4           360.0        104.43          2288.0                NaN   
5          2747.0        106.08          2321.0                NaN   
6          2833.0        106.22          2341.9                NaN   
7          1731.0        106.79          2349.6                NaN   
8          2630.0        106.53          2356.2                NaN   
9          2785.0        106.29          2321.0                NaN   

   nse_high_price  nse_low_price  ...  nse_traded_val  nse_avg_rate  \
0             NaN            NaN  ...             NaN           NaN   
1             NaN            NaN  ...             NaN           NaN   
2             NaN            NaN  ...             NaN           NaN   
3             NaN            NaN  ...             NaN           NaN   
4             NaN            NaN  ...             NaN           NaN   
5             NaN            NaN  ...             NaN           NaN   
6             NaN            NaN  ...             NaN           NaN   
7             NaN            NaN  ...             NaN           NaN   
8             NaN            NaN  ...             NaN           NaN   
9             NaN            NaN  ...             NaN           NaN   

   nse_market_cap  nse_pe  nse_pb  nse_bv_per_share  equity_shares_on_stkdate  \
0             NaN     NaN     NaN               NaN                22000000.0   
1             NaN     NaN     NaN               NaN                22000000.0   
2             NaN     NaN     NaN               NaN                22000000.0   
3             NaN     NaN     NaN               NaN                22000000.0   
4             NaN     NaN     NaN               NaN                22000000.0   
5             NaN     NaN     NaN               NaN                22000000.0   
6             NaN     NaN     NaN               NaN                22000000.0   
7             NaN     NaN     NaN               NaN                22000000.0   
8             NaN     NaN     NaN               NaN                22000000.0   
9             NaN     NaN     NaN               NaN                22000000.0   

   KZ Index        Ownership  Quartile  
0 -0.335171  Ram Ratna Group         3  
1 -0.335171  Ram Ratna Group         3  
2 -0.335171  Ram Ratna Group         3  
3 -0.335171  Ram Ratna Group         3  
4 -0.335171  Ram Ratna Group         3  
5 -0.335171  Ram Ratna Group         3  
6 -0.335171  Ram Ratna Group         3  
7 -0.335171  Ram Ratna Group         3  
8 -0.335171  Ram Ratna Group         3  
9 -0.335171  Ram Ratna Group         3  

[10 rows x 23 columns]

In [25]:
df_merged['company_name'].value_counts()

AGARWAL INDUSTRIAL CORPN. LTD.     19
VARROC ENGINEERING LTD.            19
RUSHIL DECOR LTD.                  19
SHRIRAM PISTONS & RINGS LTD.       19
INFIBEAM AVENUES LTD.              19
                                   ..
RAJSHREE POLYPACK LTD.             10
MADHYA PRADESH TODAY MEDIA LTD.    10
KRISHANA PHOSCHEM LTD.             10
DYNAMIC CABLES LTD.                 9
VISHAL BEARINGS LTD.                2
Name: company_name, Length: 227, dtype: int64

## Conclusion
- In the final data, there are 227 companies
- We have the stock data of anything between 10-19 days for each company
- Some companies have NaN values in bse columns whereas some in NSE columns, didnt know what to do with those
- Scope to drop some columns which arent required for the analysis, right now there are 23 columns
- All datatype mismatches have been corrected and date column made proper
- Added the KZ index and Quartile column from the companies.xlsx file, might be useful for analysis later

In [26]:
df_merged.to_csv('data.csv', encoding='utf-8', index=False)